File: packages/swift-client/Sources/PyricFirestore/Public/CollectionReference.swift
import Foundation

public final class CollectionReference: Query, @unchecked Sendable {

    override public var collectionID: String {
        guard let last = path.components(separatedBy: "/").last else { return "" }
        return last
    }

    public var parent: DocumentReference? {
        let segments = path.components(separatedBy: "/")
        if segments.count <= 1 { return nil }
        let parentPath = segments.dropLast().joined(separator: "/")
        return DocumentReference(firestore: firestore, path: parentPath)
    }

    public init(firestore: Firestore, path: String) {
        super.init(firestore: firestore, rootSource: .collection(path: path), path: path)
    }

    // ── Document Accessors ───────────────────────────────────────────────────

    public func document() -> DocumentReference {
        let autoId = AutoId.new()
        return DocumentReference(firestore: firestore, path: "\(path)/\(autoId)")
    }

    public func document(_ documentPath: String) -> DocumentReference {
        let clean = documentPath.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        precondition(!clean.isEmpty, "Document path must not be empty.")
        return DocumentReference(firestore: firestore, path: "\(path)/\(clean)")
    }

    // ── Document Insertion ───────────────────────────────────────────────────

    @discardableResult
    public func addDocument(data: [String: Any]) async throws -> DocumentReference {
        let docRef = document()
        try await docRef.setData(data)
        return docRef
    }

    @discardableResult
    public func addDocument(
        data: [String: Any],
        completion: (@Sendable (Error?) -> Void)? = nil
    ) -> DocumentReference {
        let docRef = document()
        docRef.setData(data, completion: completion)
        return docRef
    }
}

File: packages/kt-client/src/main/kotlin/com/google/firebase/firestore/CollectionReference.kt
package com.google.firebase.firestore

import com.google.android.gms.tasks.Task
import dev.pyric.codecs.AutoId

class CollectionReference internal constructor(
    firestore: FirebaseFirestore,
    path: String
) : Query(firestore, path = path, collectionId = path.substringAfterLast('/')) {

    val id: String get() = collectionId

    val parent: DocumentReference?
        get() = if (path != null && path.contains('/')) {
            firestore.document(path.substringBeforeLast('/'))
        } else null

    fun document(): DocumentReference = document(AutoId.generate())

    fun document(documentPath: String): DocumentReference {
        val trimmed = documentPath.trim('/')
        return firestore.document("$path/$trimmed")
    }

    fun add(data: Any): Task<DocumentReference> {
        val docRef = document()
        return docRef.set(data).continueWith { docRef }
    }
}
